File: class21/python/GQS2101-SikuliX_Conceitos.py
#!/usr/bin/env python3
"""
GQS2101 - Conceitos de automacao visual (estilo SikuliX), em Python puro.

O SikuliX real depende do runtime Jython embutido no proprio SikuliX IDE
(nao existe `pip install sikuli` para CPython - ver Aula 21, Slide 4).
Por isso, este script NAO importa a biblioteca `sikuli`: ele MODELA a
logica central do SikuliX (busca de imagem na tela por similaridade,
click, type, wait, exists) usando uma "tela simulada" (mock), para que o
aluno entenda o algoritmo sem precisar de ambiente grafico.

Uso:
    python3 GQS2101-SikuliX_Conceitos.py
"""

import time
from dataclasses import dataclass
from typing import Optional


@dataclass
class Regiao:
    """Representa a regiao da tela (real ou simulada) onde uma imagem foi encontrada."""
    x: int
    y: int
    largura: int
    altura: int
    similaridade: float

    def centro(self) -> tuple:
        return (self.x + self.largura // 2, self.y + self.altura // 2)


class Settings:
    """Equivalente conceitual de org.sikuli.script.Settings."""
    MinSimilarity = 0.7  # limiar padrao (0.0 a 1.0) - equivalente ao Settings.MinSimilarity do SikuliX


class ImageMatcher:
    """
    Simula o reconhecimento de imagem do SikuliX.

    Em vez de varrer pixels de uma captura de tela real, mantemos um
    "mapa da tela simulada": um dicionario que associa o nome do arquivo
    de imagem de referencia a uma Regiao com uma similaridade pre-calculada
    (como se um algoritmo de correlacao de template ja tivesse rodado).
    Isso reproduz o comportamento observavel de find/click/type/wait/exists
    sem exigir interface grafica.
    """

    def __init__(self, min_similarity: float = None):
        self.min_similarity = (
            min_similarity if min_similarity is not None else Settings.MinSimilarity
        )
        # "tela simulada": imagem de referencia -> (regiao, similaridade encontrada)
        self._tela: dict[str, Regiao] = {}

    def registrar_elemento(self, imagem: str, x: int, y: int, largura: int,
                            altura: int, similaridade: float) -> None:
        """Adiciona um elemento visivel na tela simulada (setup do cenario de teste)."""
        self._tela[imagem] = Regiao(x, y, largura, altura, similaridade)

    def find(self, imagem: str) -> Optional[Regiao]:
        """
        Procura a imagem na tela simulada.
        Retorna a Regiao se a similaridade encontrada for >= min_similarity,
        caso contrario retorna None (equivalente a FindFailed no SikuliX real).
        """
        regiao = self._tela.get(imagem)
        if regiao is None:
            return None
        if regiao.similaridade < self.min_similarity:
            return None
        return regiao

    def exists(self, imagem: str) -> bool:
        return self.find(imagem) is not None

    def click(self, imagem: str) -> bool:
        regiao = self.find(imagem)
        if regiao is None:
            print(f"  [click] FALHOU: '{imagem}' nao encontrada "
                  f"(similaridade minima exigida: {self.min_similarity:.2f})")
            return False
        cx, cy = regiao.centro()
        print(f"  [click] '{imagem}' encontrada em ({cx}, {cy}) "
              f"com similaridade {regiao.similaridade:.2f} -> clique simulado")
        return True

    def type(self, imagem: str, texto: str) -> bool:
        if not self.click(imagem):
            return False
        print(f"  [type] digitando '{texto}' no campo '{imagem}'")
        return True

    def wait(self, imagem: str, timeout: float, intervalo: float = 0.3) -> Optional[Regiao]:
        """
        Simula o polling do wait() do SikuliX: tenta encontrar a imagem
        repetidamente ate o timeout. Aqui nao ha "chegada tardia" real do
        elemento (tela simulada e estatica), mas o laco documenta o
        comportamento esperado do SikuliX real.
        """
        decorrido = 0.0
        while decorrido < timeout:
            regiao = self.find(imagem)
            if regiao is not None:
                print(f"  [wait] '{imagem}' apareceu apos {decorrido:.1f}s")
                return regiao
            time.sleep(intervalo)
            decorrido += intervalo
        print(f"  [wait] TIMEOUT: '{imagem}' nao apareceu em {timeout}s")
        return None


def demo_login_calculadora():
    """Demonstra um fluxo tipico: login simulado + calculo 12 + 34 na calculadora."""
    matcher = ImageMatcher(min_similarity=0.8)

    # Cenario 1: tela de login com boa qualidade de imagem (similaridade alta)
    matcher.registrar_elemento("campo-usuario.png", 100, 200, 150, 30, similaridade=0.95)
    matcher.registrar_elemento("campo-senha.png", 100, 250, 150, 30, similaridade=0.93)
    matcher.registrar_elemento("btn-login.png", 100, 300, 80, 30, similaridade=0.90)
    matcher.registrar_elemento("msg-boas-vindas.png", 100, 350, 200, 20, similaridade=0.85)

    print("=" * 60)
    print("  CENARIO 1: Login (imagens com boa similaridade)")
    print("=" * 60)
    matcher.type("campo-usuario.png", "admin")
    matcher.type("campo-senha.png", "senha123")
    matcher.click("btn-login.png")
    if matcher.exists("msg-boas-vindas.png"):
        print("  Login OK!")
    else:
        print("  Falha no login")

    # Cenario 2: calculadora - imagem de um botao degradada (baixa similaridade,
    # por exemplo por causa de DPI diferente do usado na captura original)
    print()
    print("=" * 60)
    print("  CENARIO 2: Calculadora 12 + 34 (botao '=' com baixa similaridade)")
    print("=" * 60)
    matcher2 = ImageMatcher(min_similarity=Settings.MinSimilarity)
    matcher2.registrar_elemento("btn-1.png", 10, 10, 40, 40, similaridade=0.92)
    matcher2.registrar_elemento("btn-2.png", 60, 10, 40, 40, similaridade=0.91)
    matcher2.registrar_elemento("btn-3.png", 110, 10, 40, 40, similaridade=0.90)
    matcher2.registrar_elemento("btn-4.png", 10, 60, 40, 40, similaridade=0.88)
    matcher2.registrar_elemento("btn-plus.png", 160, 10, 40, 40, similaridade=0.85)
    # Simula um botao "=" cuja imagem de referencia foi capturada em outra
    # resolucao de tela: a similaridade cai abaixo do MinSimilarity padrao.
    matcher2.registrar_elemento("btn-igual.png", 160, 110, 40, 40, similaridade=0.55)

    matcher2.click("btn-1.png")
    matcher2.click("btn-2.png")
    matcher2.click("btn-plus.png")
    matcher2.click("btn-3.png")
    matcher2.click("btn-4.png")
    sucesso = matcher2.click("btn-igual.png")

    if not sucesso:
        print("\n  Mitigacao: reduzir Settings.MinSimilarity ou recapturar a imagem")
        print("  'btn-igual.png' na resolucao/DPI atual do ambiente de teste.")
        matcher2.min_similarity = 0.5
        print(f"  Novo MinSimilarity = {matcher2.min_similarity:.2f}")
        matcher2.click("btn-igual.png")

    # Cenario 3: wait() por um elemento ausente (demonstra timeout)
    print()
    print("=" * 60)
    print("  CENARIO 3: wait() por elemento que nunca aparece")
    print("=" * 60)
    matcher3 = ImageMatcher()
    matcher3.wait("dialogo-carregamento-fim.png", timeout=1.0, intervalo=0.3)


def main():
    print("GQS2101 - Conceitos de Automacao Visual (modelo do SikuliX)")
    demo_login_calculadora()
    print("\nExecucao concluida sem erros.")


if __name__ == "__main__":
    main()

File: class21/java/GQS2101-SikuliX_Conceitos.java
/**
 * GQS2101 - Conceitos de automacao visual (estilo SikuliX), em Java puro.
 *
 * O SikuliX real (org.sikuli.script.*) exige um ambiente grafico e a
 * biblioteca nativa de reconhecimento de imagem (OpenCV). Este arquivo NAO
 * depende do jar do SikuliX: ele MODELA a logica central da ferramenta
 * (busca de imagem na tela por similaridade, click, type, wait, exists)
 * usando uma "tela simulada" (mock), para fins didaticos.
 *
 * Compilacao e execucao:
 *   javac GQS2101-SikuliX_Conceitos.java
 *   java -cp . GQS2101_SikuliX_Conceitos
 */
import java.util.HashMap;
import java.util.Map;

class Regiao {
    int x, y, largura, altura;
    double similaridade;

    Regiao(int x, int y, int largura, int altura, double similaridade) {
        this.x = x;
        this.y = y;
        this.largura = largura;
        this.altura = altura;
        this.similaridade = similaridade;
    }

    int centroX() {
        return x + largura / 2;
    }

    int centroY() {
        return y + altura / 2;
    }
}

/** Equivalente conceitual de org.sikuli.script.Settings. */
class Settings {
    static double MinSimilarity = 0.7;
}

class ImageMatcher {

    private double minSimilarity;
    private final Map<String, Regiao> tela = new HashMap<>();

    ImageMatcher() {
        this(Settings.MinSimilarity);
    }

    ImageMatcher(double minSimilarity) {
        this.minSimilarity = minSimilarity;
    }

    void setMinSimilarity(double valor) {
        this.minSimilarity = valor;
    }

    double getMinSimilarity() {
        return minSimilarity;
    }

    void registrarElemento(String imagem, int x, int y, int largura, int altura, double similaridade) {
        tela.put(imagem, new Regiao(x, y, largura, altura, similaridade));
    }

    Regiao find(String imagem) {
        Regiao regiao = tela.get(imagem);
        if (regiao == null || regiao.similaridade < minSimilarity) {
            return null;
        }
        return regiao;
    }

    boolean exists(String imagem) {
        return find(imagem) != null;
    }

    boolean click(String imagem) {
        Regiao regiao = find(imagem);
        if (regiao == null) {
            System.out.printf("  [click] FALHOU: '%s' nao encontrada (similaridade minima exigida: %.2f)%n",
                imagem, minSimilarity);
            return false;
        }
        System.out.printf("  [click] '%s' encontrada em (%d, %d) com similaridade %.2f -> clique simulado%n",
            imagem, regiao.centroX(), regiao.centroY(), regiao.similaridade);
        return true;
    }

    boolean type(String imagem, String texto) {
        if (!click(imagem)) {
            return false;
        }
        System.out.printf("  [type] digitando '%s' no campo '%s'%n", texto, imagem);
        return true;
    }

    Regiao wait(String imagem, double timeoutSegundos, double intervaloSegundos) {
        double decorrido = 0.0;
        while (decorrido < timeoutSegundos) {
            Regiao regiao = find(imagem);
            if (regiao != null) {
                System.out.printf("  [wait] '%s' apareceu apos %.1fs%n", imagem, decorrido);
                return regiao;
            }
            try {
                Thread.sleep((long) (intervaloSegundos * 1000));
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            }
            decorrido += intervaloSegundos;
        }
        System.out.printf("  [wait] TIMEOUT: '%s' nao apareceu em %.1fs%n", imagem, timeoutSegundos);
        return null;
    }
}

class GQS2101_SikuliX_Conceitos {

    static void demoLoginCalculadora() {
        ImageMatcher matcher = new ImageMatcher(0.8);

        System.out.println("=".repeat(60));
        System.out.println("  CENARIO 1: Login (imagens com boa similaridade)");
        System.out.println("=".repeat(60));
        matcher.registrarElemento("campo-usuario.png", 100, 200, 150, 30, 0.95);
        matcher.registrarElemento("campo-senha.png", 100, 250, 150, 30, 0.93);
        matcher.registrarElemento("btn-login.png", 100, 300, 80, 30, 0.90);
        matcher.registrarElemento("msg-boas-vindas.png", 100, 350, 200, 20, 0.85);

        matcher.type("campo-usuario.png", "admin");
        matcher.type("campo-senha.png", "senha123");
        matcher.click("btn-login.png");
        System.out.println(matcher.exists("msg-boas-vindas.png") ? "  Login OK!" : "  Falha no login");

        System.out.println();
        System.out.println("=".repeat(60));
        System.out.println("  CENARIO 2: Calculadora 12 + 34 (botao '=' com baixa similaridade)");
        System.out.println("=".repeat(60));
        ImageMatcher matcher2 = new ImageMatcher(Settings.MinSimilarity);
        matcher2.registrarElemento("btn-1.png", 10, 10, 40, 40, 0.92);
        matcher2.registrarElemento("btn-2.png", 60, 10, 40, 40, 0.91);
        matcher2.registrarElemento("btn-3.png", 110, 10, 40, 40, 0.90);
        matcher2.registrarElemento("btn-4.png", 10, 60, 40, 40, 0.88);
        matcher2.registrarElemento("btn-plus.png", 160, 10, 40, 40, 0.85);
        // Botao "=" capturado em outra resolucao/DPI: similaridade abaixo do padrao
        matcher2.registrarElemento("btn-igual.png", 160, 110, 40, 40, 0.55);

        matcher2.click("btn-1.png");
        matcher2.click("btn-2.png");
        matcher2.click("btn-plus.png");
        matcher2.click("btn-3.png");
        matcher2.click("btn-4.png");
        boolean sucesso = matcher2.click("btn-igual.png");

        if (!sucesso) {
            System.out.println();
            System.out.println("  Mitigacao: reduzir Settings.MinSimilarity ou recapturar a imagem");
            System.out.println("  'btn-igual.png' na resolucao/DPI atual do ambiente de teste.");
            matcher2.setMinSimilarity(0.5);
            System.out.printf("  Novo MinSimilarity = %.2f%n", matcher2.getMinSimilarity());
            matcher2.click("btn-igual.png");
        }

        System.out.println();
        System.out.println("=".repeat(60));
        System.out.println("  CENARIO 3: wait() por elemento que nunca aparece");
        System.out.println("=".repeat(60));
        ImageMatcher matcher3 = new ImageMatcher();
        matcher3.wait("dialogo-carregamento-fim.png", 1.0, 0.3);
    }

    public static void main(String[] args) {
        System.out.println("GQS2101 - Conceitos de Automacao Visual (modelo do SikuliX)");
        demoLoginCalculadora();
        System.out.println("\nExecucao concluida sem erros.");
    }
}
